src/arcsf/config/experiment.py

```diff
@@ -0,0 +1,304 @@
+import os
+import warnings
+from copy import copy
+from itertools import product
+
+import wandb
+import yaml
+from jinja2 import Environment, FileSystemLoader
+
+import arcsf.constants
+from arcsf.config.config import Config
+from arcsf.data.config import DataConfig
+from arcsf.models.config import ModelConfig
+
+
+def _listify(obj: object):
+    """Takes an object. If a list, returns it. If not, returns a list containing
+    the object.
+
+    Args:
+        obj: Object to ensure is a list.
+
+    Returns:
+        Either the original object if a list or a list containing the object.
+    """
+    if isinstance(obj, list):
+        return obj
+    return [obj]
+
+
+def _make_config_name(
+    top_config_name: str,
+    n: int,
+) -> str:
+    return f"{top_config_name}/experiment_{n}"
+
+
+def _generate_combos_from_dict(
+    combo_dict: dict,
+    full: bool,
+) -> dict:
+    # Convert combo dict into dict of combos
+    sweep_dict_keys, sweep_dict_vals = zip(*combo_dict.items())
+    combinations = [
+        dict(zip(sweep_dict_keys, v)) for v in product(*list(sweep_dict_vals))
+    ]
+
+    # Initialise train_type
+    train_type = "retain"
+    if full:
+        train_type = "full"
+
+    # Add train_type to each combination
+    for n, _ in enumerate(combinations):
+        combinations[n]["train_type"] = train_type
+
+    # Return
+    return combinations
+
+
+def generate_combos_from_dict(
+    combo_dict: dict,
+    full: bool,
+    wandb_kwargs: dict,
+) -> dict:
+    # Generate combinations
+    combinations = _generate_combos_from_dict(combo_dict, full)
+
+    # Process model kwargs
+    for n, _ in enumerate(combinations):
+        combinations[n]["hyperparameter_config"] = combinations[n]["model_config"][1]
+        combinations[n]["model_config"] = combinations[n]["model_config"][0]
+        combinations[n] = {**combinations[n], **wandb_kwargs}
+
```

Comment:
Why there's `["model_config"][1]` and `["model_config"][0]` confused me. I see where it's coming from now but a comment here would be helpful (about how model + params are specified in the top level config), maybe also think if there's a better name for it in the configs e.g. `model_and_params` though that's pretty long and clunky.

```suggestion
    for n, comb in enumerate(combinations):
        # Top level model_config contains [model, hyperparameters] configs
        comb["model_config"] = comb["model_config"][0]
        comb["hyperparameter_config"] = comb["model_config"][1]
        combinations[n] = {**comb, **wandb_kwargs}

```

Reply:
Have adopted your suggestion